File: frontend/src/services/resume.ts
import apiClient from './api';
import type { ApiResponse } from '../types/api';
import type {
  ResumeSubmitRequest,
  ResumeParseResponse,
  ResumeSubmitResponse,
  ResumeUploadURLResponse,
} from '../types/resume';

const DEFAULT_PDF_CONTENT_TYPE = 'application/pdf';

// 解析简历 PDF：先拿预签名地址，再直传对象存储，最后用 object_key 触发解析
export const parseResume = async (file: File): Promise<ResumeParseResponse> => {
  const uploadURLResponse = await apiClient.get<ApiResponse<ResumeUploadURLResponse>>(
    '/resume/upload-url',
    {
      params: {
        filename: file.name,
      },
    }
  );

  const { upload_url, object_key } = uploadURLResponse.data.data;

  const uploadResponse = await fetch(upload_url, {
    method: 'PUT',
    headers: {
      'Content-Type': file.type || DEFAULT_PDF_CONTENT_TYPE,
    },
    body: file,
  });

  if (!uploadResponse.ok) {
    throw new Error(`简历上传失败 (${uploadResponse.status})`);
  }

  const response = await apiClient.post<ApiResponse<ResumeParseResponse>>(
    '/resume/parse',
    {
      object_key,
    }
  );

  return response.data.data;
};

// 提交简历信息
export const submitResume = async (data: ResumeSubmitRequest): Promise<ResumeSubmitResponse> => {
  const response = await apiClient.post<ApiResponse<ResumeSubmitResponse>>('/resume/submit', data);
  return response.data.data;
};

File: internal/service/resume.go
package service

import (
	"context"

	"ai_interview/internal/domain"
)

// ResumeService 简历模块业务逻辑接口
type ResumeService interface {
	// PresignUpload 生成简历 PDF 直传 S3 的预签名 PUT URL（5 分钟有效）。
	// 返回 uploadURL（前端 PUT 到此地址）和 objectKey（后续 parse 时传回）。
	PresignUpload(ctx context.Context, userID, filename string) (uploadURL, objectKey string, err error)

	// Parse 解析 PDF 简历，同步返回结构化数据。
	// objectKey 为前端直传后 S3 中的路径，由 PresignUpload 返回。
	Parse(ctx context.Context, userID, objectKey string) (*domain.StructuredResume, error)

	// Submit 保存用户确认后的简历，存入 PG 主存储 + Redis 1h 缓存。
	Submit(ctx context.Context, resume domain.StructuredResume) (resumeID string, err error)

	// Get 查询用户当前简历（Redis → PG 回填）。
	Get(ctx context.Context, userID string) (*domain.StructuredResume, error)
}
